hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/update/server/SCMUpdateClientManager.java

```diff
@@ -0,0 +1,149 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package org.apache.hadoop.hdds.scm.update.server;
+
+import org.apache.hadoop.hdds.scm.exceptions.SCMException;
+import org.apache.ratis.thirdparty.io.grpc.Status;
+import org.apache.ratis.thirdparty.io.grpc.stub.StreamObserver;
+import org.apache.hadoop.hdds.protocol.scm.proto.SCMUpdateServiceProtos.Type;
+import org.apache.hadoop.hdds.protocol.scm.proto.SCMUpdateServiceProtos.UpdateRequest;
+import org.apache.hadoop.hdds.protocol.scm.proto.SCMUpdateServiceProtos.UpdateResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Class that manages SCM update clients.
+ */
+public class SCMUpdateClientManager {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(SCMUpdateClientManager.class);
+  private Map<UUID, SCMUpdateClientInfo> clients;
+  private Map<Type, SCMUpdateHandler> handlers;
+
+  public SCMUpdateClientManager() {
+    clients = new ConcurrentHashMap<>();
+    handlers = new ConcurrentHashMap<>();
+  }
+
+  public void registerHandler(SCMUpdateHandler handler) {
+    handlers.put(handler.getType(), handler);
+  }
+
+  public void unRegisterHandler(Type type) {
+    handlers.remove(type);
+  }
+
+  public UUID addClient() throws SCMException {
+    UUID clientId = UUID.randomUUID();
+    int retryCount = 5;
+    while (clients.containsKey(clientId)) {
+      if (retryCount > 0) {
+        clientId = UUID.randomUUID();
+        retryCount--;
+      } else {
+        throw new SCMException("Failed to add CRL client with random clientId" +
+            " collision", SCMException.ResultCodes.FAILED_TO_ADD_CRL_CLIENT);
+      }
+    }
+
+    SCMUpdateClientInfo clientInfo = new SCMUpdateClientInfo(clientId);
+    clients.put(clientId, clientInfo);
+    return clientId;
+  }
+
+  // this does not necessarily produce a server response via responseObserver.
+  public void handleClientUpdate(UpdateRequest request,
+      StreamObserver<UpdateResponse> responseObserver) {
+    UUID clientId = SCMUpdateClientInfo.fromClientIdProto(
+        request.getClientId());
+
+    // Unknown client update
+    if (!clients.containsKey(clientId)) {
+      responseObserver.onError(Status.INVALID_ARGUMENT
+          .withDescription("Client must subscribe before it can " +
+              "send/receive updates")
+          .asException());
+    }
+
+    // record the server to client channel
+    SCMUpdateClientInfo clientInfo = clients.get(clientId);
+    if (clientInfo.getResponseObserver() == null) {
+      clientInfo.setResponseObserver(responseObserver);
+    }
+
+    if (handlers.containsKey(request.getUpdateType())) {
+      handlers.get(request.getUpdateType())
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `get()` may not be thread-safe at lines: **94 and 95**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The remove that can remove the key is at line: **53**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).